### ISSUE TITLE
Large memory spike after introduction of histogram size bound

### ISSUE THREAD

**Issue by fjetter** [dmlc/xgboost#9452]
https://github.com/dmlc/xgboost/pull/9440 introduced a bound to histogram sizes but the behavior is not entirely what I would expect.

For small-ish values of `max_depth` (I tested 12) I am indeed observing a reduction of memory usage and an improvement in end to end runtime performance when running on dask (looks like network is much more consistently utilized which is nice). Good job!

However, when dialing up the depth of the tree, I am actually observing a higher memory usage than before.

The below plots show cluster memory usage when running on a dask cluster with four workers.

Stable (xgboost 1.7.6)

![image](https://github.com/dmlc/xgboost/assets/8629629/c97450fe-201f-4a28-befa-6d940f98f86f)

Nightly (merge commit of https://github.com/dmlc/xgboost/pull/9440)

![image](https://github.com/dmlc/xgboost/assets/8629629/25f54dd2-3573-490e-b662-3f864f2428e9)

Not only is the new version exhibiting a very steep spike that is higher than what is on stable, the usage also seems to grow with increasing number of boosting rounds. On larger examples I could observe this to be even more severe.

To generate these plots I was using the same code as provided in  but was running the examples on Coiled. Similar examples should work when running on a LocalCluster

```python
import dask
import xgboost
from dask.distributed import Client
from distributed.diagnostics import MemorySampler

cluster = coiled.Cluster(
    n_workers=4,
    worker_vm_types=["r6i.xlarge"]
)
client = Client(cluster)

ms = MemorySampler()

dtypes = {}

for i in range(40):
    dtypes["f-%d" % i] = float
    
for i in range(40):
    dtypes["i-%d" % i] = int
    
df = dask.datasets.timeseries(
    start="2000-01-01",
    end="2001-01-01",
    dtypes=dtypes,
    partition_freq="4d",
    freq="10s",
)

df = df.persist()  # this is a 1.9GiB dataset 
X = df.drop("i-1", axis=1)
y = df["i-1"]

d_train = xgboost.dask.DaskDMatrix(
    None, X, y, enable_categorical=True
)

for num_boost_round in [5, 10, 15]:
    for max_depth in [18]:
        print(f"{max_depth=}_{num_boost_round=}")
        with ms.sample(f"stable-{max_depth=}_{num_boost_round=}"):
            model = xgboost.dask.train(
                None,
                {
                    "max_depth": max_depth, 
                },
                d_train,
                num_boost_round=num_boost_round,
                evals=[(d_train, "train")],
            )
            
ms.plot(align=True, grid=True)

```

xref https://github.com/dmlc/xgboost/issues/9342 https://github.com/dmlc/xgboost/pull/9440

**Comment by trivialfis** [comment]
Thank you for raising the issue and sharing the script! I'm trying to profile it, the spike seems to be introduced eailer than the PR. Will keep you posted.

**Comment by trivialfis** [comment]
Note: The accuracy starts off significantly better on master. `train-rmse:31.18094` on master while `train-rmse:700.37133` on 1.7.6. I think this is contributed by the initial estimation of intercept. With this, it's not unexpected for XGBoost to grow deeper trees.

**Comment by trivialfis** [comment]
Opened a PR https://github.com/dmlc/xgboost/pull/9455 for mitigation.

**Comment by fjetter** [comment]
Thanks for the quick turnaround.

I ran a couple of experiments with cache sizes to get a feeling for the runtime implications on smaller trees. And ran into something that is very unintuitive.
I ran a training with `max_depth=12` with a cache size of `2**15` and `2**12` to confirm the null hypothesis. As expected, both are large enough for the cache to not impact anything. When reducing the size of the cache, I expected the runtime to increase but was experiencing the opposite effect.

Smaller cache sizes actually terminated the training earlier. The plot below shows log2(size) of 11 and 10 and the runtime does not really make sense to me. I only ran 3 boost iterations here.

![image](https://github.com/dmlc/xgboost/assets/8629629/4b974873-2070-4c88-bbac-9e2310b1c987)
